app/assets/javascripts/editor.js

```diff
@@ -104,16 +104,74 @@ class Editor {
   // h7() {
   //   this.wrap('#######','')
   // }
+
+  //debounce function addition
+  debounce(func, wait, immediate) {
+    let timeout;
+    return function () {
+      let context = this,
+        args = arguments;
+      let later = function () {
+        timeout = null;
+        if (!immediate) func.apply(context, args);
+      };
+      let callNow = immediate && !timeout;
+      clearTimeout(timeout);
```

Comment:
Why, clear timeout here, if the variable has a null value in line 115, why not just reassign it on line 120 and skip doing line 119?

Reply:
@RuthNjeri, at line 115 - 
```javascript
timeout = null;
```
the timeout variable gets cleared to make the next execution run as soon as the function is called, i.e. at line 120:
```javascript
      if (callNow) func.apply(context, args);
```

On the other hand, `clearTimeout` at line 119:
```javascript
clearTimeout(timeout);
```
 is used to clear out the previous created timeout, so clearTimeout should be done before setTimeout at line 120: 
```javascript
      timeout = setTimeout(later, wait);
```
in order to cancel out previous invocation of setTimeout function. Thus, `clearTimeout` is essential for the invocation the the `setTimeout` function 😅...